### PR TITLE
[Example] Add Top-n-sigma logit truncation custom logits processor

### LINKED ISSUE #45023: [RFC] Add Top-nσ logit truncation example via custom logits processor
## RFC: Add Top-nσ Logit Truncation Example via Custom Logits Processor

**Reference**: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### Motivation

Top-nσ is a logit-space dynamic truncation method proposed by Tang et al. (ACL 2025). Unlike probability-space filters such as `top_p` or `min_p`, it applies **before softmax** and keeps the candidate set less sensitive to temperature changes.

This issue proposes adding a **research-oriented example implementation** using vLLM's existing custom logits processor API. It is **not** a request to add a new built-in sampling parameter.

### Scope

This proposal only adds an **example file** demonstrating how to implement top-nσ as a custom logits processor. It:
- Does **not** add a new `SamplingParams` field
- Does **not** change any existing sampling behavior
- Does **not** modify the core `Sampler` class

### Algorithm

For each request row:

```
1. max_logit = logits.max()
2. std_logit = logits.std()
3. threshold = max_logit - n * std_logit
4. logits[logits < threshold] = -inf   (mask unlikely tokens)
5. Let existing vLLM sampler handle temperature, softmax, and sampling
```

The standard deviation directly measures the "peakiness" of the logit distribution:
- **Sharp distribution** (model is certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model is uncertain): large std → wide threshold → more candidates

### How it compares to other truncation methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative probability ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |
| temperature | Logit | logits / T — rescales distribution steepness | ❌ No truncation |

> Key distinction: top_k/top_p/min_p operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space).

### Why Top-nσ?

- **Truncation before softmax** — top_nσ filters in logit space first, then softmax only processes the surviving candidates (~2.5% of vocab). In contrast, top_p/min_p must compute a full-vocab softmax first, then truncate in probability space.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula: sharp distributions auto-narrow, flat ones auto-widen. top_k fixes the count; top_p adapts the count but not the threshold tightness.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally), while top_p/min_p candidate counts vary significantly with T.
- **Argmax-invariant** — the top-1 token always survives (0 std from itself), so greedy decoding can skip this processor entirely.

### Usage (via custom logits processor)

```python
from vllm import SamplingParams

sampling_params = SamplingParams(
    temperature=0.8,
    extra_args={"top_n_sigma": 2.0}
)
```

The custom logit processor reads `top_n_sigma` from `extra_args` and applies truncation before the standard sampling pipeline.

### Safety / Edge Cases

| Case | Behavior |
|------|----------|
| `n <= 0` | Reject (raise `ValueError`) |
| `std == 0` (all logits equal) | Keep all tokens (no filtering) |
| NaN/Inf logits | Preserve existing behavior (no-op) |
| After filtering | **At least the argmax token is always kept** (prevent all-`-inf`) |

### Comparison Example

```python
# Logit distribution: max=10.0, std=2.0

top_k=5:           Keeps exactly 5 tokens
top_p=0.9:         Keeps tokens summing to 90% cumulative probability
min_p=0.05:        Keeps tokens with prob >= 5% of max
top_n_sigma=2.0:   Keeps tokens with logit >= 10.0 - 2*2.0 = 6.0
```

### Deliverables (if maintainers are interested)

1. **Example file**: `examples/features/logits_processor/top_n_sigma.py`
2. **Unit tests**: Verify masking, edge cases, and compatibility with other samplers
3. **Documentation**: One-paragraph note in sampling docs

---

**References**: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### PR DESCRIPTION
**Pull request by Jackie2049**
## Summary

Closes #45023

Adds an example demonstrating **Top-nσ** as a custom logits processor in vLLM, based on the paper:

> Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### What is Top-nσ?

Top-nσ is a logit-space dynamic truncation method that uses the standard deviation of the logit distribution to set a filtering threshold:

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike probability-space filters (top_p, min_p), it operates **before softmax** and adapts to distribution "peakiness":
- **Sharp distribution** (model certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model uncertain): large std → wide threshold → more candidates

### How it compares to other truncation methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative probability ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |
| temperature | Logit | logits / T — rescales distribution steepness | ❌ No truncation |

> Key distinction: top_k/top_p/min_p operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space).

### Why Top-nσ?

- **Truncation before softmax** — top_nσ filters in logit space first, then softmax only processes the surviving candidates (~2.5% of vocab). In contrast, top_p/min_p must compute a full-vocab softmax first, then truncate in probability space.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula: sharp distributions auto-narrow, flat ones auto-widen. top_k fixes the count; top_p adapts the count but not the threshold tightness.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally), while top_p/min_p candidate counts vary significantly with T.
- **Argmax-invariant** — the top-1 token always survives (0 std from itself), so greedy decoding can skip this processor entirely.

## Changes

- New file: `examples/features/logits_processor/top_n_sigma.py`
- `apply()` method uses vectorized Torch operations (not Python per-row loop)

### Key properties
- **Argmax-invariant**: doesn't affect greedy decoding
- Uses vLLM's `extra_args` mechanism (no `SamplingParams` changes needed)
- Handles edge cases: NaN/Inf logits, zero std, n≤0 validation

### Usage

```python
from vllm import LLM, SamplingParams

llm = LLM(
    model="facebook/opt-125m",
    logits_processors=[TopNSigmaLogitsProcessor],
)

sampling_params = SamplingParams(
    temperature=0.8,
    extra_args={"top_n_sigma": 2.0}
)
```

## Testing

- [x] Example runs with `facebook/opt-125m`
- [x] Vectorized `apply()` is bit-for-bit identical to row-by-row loop across FP32/FP16 configs
- [x] Argmax-invariant verified across 32 configurations (all dtype/sharpness/n combinations)
- [x] Edge cases handled: NaN/Inf logits, std=0, n≤0 validation, empty `req_info`

### CODE REVIEW

**Comment by Codekiing** [comment]
Hi maintainers,

I’d like to kindly follow up on this PR.

It has been a few days, and the PR is still blocked before CI by the first-time contributor verification gate:

`PR must have the 'verified' or 'ready' label, or the author must have at least 4 merged PRs.`

This PR is intended to be a self-contained example under `examples/features/logits_processor/`, demonstrating Top-nσ logit truncation via the existing custom logits processor API. It does not add a new `SamplingParams` field, does not change existing sampling behavior, and does not modify the core sampler.

If the scope or placement looks unsuitable, I’m very happy to adjust it, move it into docs, or simplify it further.

Could someone with maintainer permissions help apply the appropriate `ready` / `verified` label so CI can run?

cc @simon-mo 

Thanks a lot!

**Comment by Hellisotherpeople** [comment]
Why isn't this merged yesterday? This is currently one of the SOTA samplers and is better than min_p (I'm one of the min_p authors, so I am an authority on this topic). 

Merge it now! @simon-mo

**Comment by Codekiing** [comment]
> Why isn't this merged yesterday? This is currently one of the SOTA samplers and is better than min_p (I'm one of the min_p authors, so I am an authority on this topic).
> 
> Merge it now! @simon-mo

Thanks a lot @Hellisotherpeople for your attention and endorsement. Glad to see distribution-aware sampling methods getting attention!
We'd welcome any review suggestions on the implementation. Feel free to give suggestions or contribute code to this PR if you find anything that can be optimized.
@simon-mo This PR is a self-contained example (1 file, no framework changes) demonstrating Top-nσ via the existing custom logit processor API. We're ready to address any review feedback promptly from maintainers.

**Review by Codekiing — COMMENTED** [review]
ok
